source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,467 @@
+can not provide.
+
+.. code-block:: java
+
+    public Optional<HealthData> fill(DataContainer container, HealthData healthData) {
+        healthData.set(Keys.MAX_HEALTH, DataUtil.getData(container, Keys.MAX_HEALTH));
+        healthData.set(Keys.HEALTH, DataUtil.getData(container, Keys.HEALTH));
+        return Optional.of(healthData);
+    }
+
+Other Methods
+~~~~~~~~~~~~~
+
+Depending on the abstract superclass used, some other methods may be required. For instance,
+``AbstractEntityDataProcessor`` needs to create ``DataManipulator`` instances in various points. It can't do this
+since it knows neither the implementation class nor the constructor to use. Therefore it utilizes an abstract
+function that has to be provided by the final implementation. This does nothing more than create a
+``DataManipulator`` with default data.
+
+If you implemented your ``DataManipulator`` as recommended, you can just use the no-args constructor.
+
+.. code-block:: java
+
+    protected HealthData createManipulator() {
+        return new SpongeHealthData();
+    }
+
+
+5. Implement the ValueProcessors
+================================
+
+Not only a ``DataManipulator`` may be offered to a ``DataHolder``, but also a keyed ``Value`` on its own.
+Therefore, you need to provide at least one ``ValueProcessor`` for every ``Key`` present in your
+``DataManipulator``. A ``ValueProcessor`` is named after the constant name of its ``Key`` in the ``Keys`` class
+in a fashion similar to its ``DataQuery``. The constant name is stripped of underscores, used in upper camel case
+and then suffixed with ``ValueProcessor``.
+
+A ``ValueProcessor`` should always inherit from ``AbstractSpongeValueProcessor``, which already will handle a
+portion of the ``supports()`` checks based on the type of the ``DataHolder``. For ``Keys.HEALTH``, we'll create
+and construct ``HealthValueProcessor`` as follows.
+
+.. code-block:: java
+
+    public class HealthValueProcessor extends AbstractSpongeValueProcessor<EntityLivingBase, Double,
+        MutableBoundedValue<Double> {
+
+        public HealthValueProcessor() {
+            super(EntityLivingBase.class, Keys.HEALTH);
+        }
+
+        [...]
+    }
+
+Now the ``AbstractSpongeValueProcessor`` will relieve us of the necessity to check if the value is supported.
+It is assumed to be supported if the target ``ValueContainer`` is of the type ``EntityLivingBase``.
+
+.. tip::
+
+    For a more fine-grained control over what ``EntityLivingBase`` objects are supported, the
+    ``supports(EntityLivingBase)`` method can be overridden.
+
+Again, most work is done by the abstraction class. We just need to implement two helper methods for creating
+a ``Value`` and its immutable counterpart and three methods to get, set and remove data.
+
+.. code-block:: java
+
+    protected MutableBoundedValue<Double> constructValue(Double value) {
+        return new SpongeBoundedValue<>(Keys.HEALTH, 20D,
```

Comment:
Order of arguments is already explained once in the document. Added internal reference. 

This will probably change again when the `SpongeValueBuilder` is created.